### PR TITLE
Docker Repository initialization

### PR DESCRIPTION
**Pull request by nupplaphil**
Initializing the Docker-Repository

- Added Friendica 3.6 stable images (apache, fpm, fpm-alpine)
- Added Friendica develop images (apache, fpm, fpm-alpine)
- Added .travis.yml with tests for the [official-images repository](https://github.com/docker-library/official-images)

Two thoughts about it:
- Travis currently failed because of the [override-cmd test](https://github.com/docker-library/official-images/blob/master/test/tests/override-cmd/run.sh), which works for me locally
- There is no "standalone" image because a combination of Friendica and MariaDB-Server is not that easy (to have a clean solution)

@friendica can you please add this repo to https://travis-ci.org for testing :-)

### CODE REVIEW

File: develop/apache/bin/friendica [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    # We want to update a given installation
+    # if there is no installation, exit
+    return
+  fi
+
+  copy_sources
+
+  echo "Upgrading Friendica"
+  if [ "$FRIENDICA_VERSION" = "develop" ]; then
+    composer "update"
```
Comment (MrPetovan) [inline develop/apache/bin/friendica]: For an update, the Composer command is actually the same as for a first install, e.g. `composer install`. Only developers get to run `composer update` once to add a new dependency or upgrade an existing one.

File: develop/fpm-alpine/bin/friendica [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    # We want to update a given installation
+    # if there is no installation, exit
+    return
+  fi
+
+  copy_sources
+
+  echo "Upgrading Friendica"
+  if [ "$FRIENDICA_VERSION" = "develop" ]; then
+    composer "update"
```
Comment (MrPetovan) [inline develop/fpm-alpine/bin/friendica]: Should be `composer install`, see previous comment about it.

File: develop/fpm/bin/friendica [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    # We want to update a given installation
+    # if there is no installation, exit
+    return
+  fi
+
+  copy_sources
+
+  echo "Upgrading Friendica"
+  if [ "$FRIENDICA_VERSION" = "develop" ]; then
+    composer "update"
```
Comment (MrPetovan) [inline develop/fpm/bin/friendica]: Should be `composer install`, see previous comment about it.

File: stable/apache/bin/friendica [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    # We want to update a given installation
+    # if there is no installation, exit
+    return
+  fi
+
+  copy_sources
+
+  echo "Upgrading Friendica"
+  if [ "$FRIENDICA_VERSION" = "develop" ]; then
+    composer "update"
```
Comment (MrPetovan) [inline stable/apache/bin/friendica]: Should be `composer install`, see previous comment about it.

File: stable/fpm-alpine/bin/friendica [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    # We want to update a given installation
+    # if there is no installation, exit
+    return
+  fi
+
+  copy_sources
+
+  echo "Upgrading Friendica"
+  if [ "$FRIENDICA_VERSION" = "develop" ]; then
+    composer "update"
```
Comment (MrPetovan) [inline stable/fpm-alpine/bin/friendica]: Should be `composer install`, see previous comment about it.

File: stable/fpm/bin/friendica [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    # We want to update a given installation
+    # if there is no installation, exit
+    return
+  fi
+
+  copy_sources
+
+  echo "Upgrading Friendica"
+  if [ "$FRIENDICA_VERSION" = "develop" ]; then
+    composer "update"
```
Comment (MrPetovan) [inline stable/fpm/bin/friendica]: Should be `composer install`, see previous comment about it.

File: README.md [hunk 7]
```diff
@@ -1,3 +1,82 @@
...
+This repository holds the official Docker Image for [Friendica](https://friendi.ca)
+
+# What is Friendica?
+
+Friendica is a decentralised communications platform that integrates social communication. Our platform links to independent social projects and corporate services.
+
+![logo](https://cdn.rawgit.com/nupplaphil/friendica-docker/c59f235f/friendica.svg)
+
+# How to use this image
+The images are designed to be used in a micro-service environment. There are two types of the image you can choose from.
```
Comment (tobiasd) [inline README.md]: Can you please stick to the general documentation rule of Friendica to only put one sentence per line. It makes finding endless sentences more easy and the diffs between textual changes get easier to read.

File: develop/fpm-alpine/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    \
+    rm /var/spool/cron/crontabs/root; \
+    echo '*/10 * * * * cd /var/www/html && php -f bin/worker.php' > /var/spool/cron/crontabs/www-data
+
+# install the PHP extensions we need
+# see https://friendi.ca/resources/requirements/
+RUN set -ex; \
+    \
+    apk add -U --no-cache --virtual .build-deps \
+        libxml2-dev \
```
Comment (tobiasd) [inline develop/fpm-alpine/Dockerfile]: I don't see an email system in the list, how do the users get their notification emails?

**Comment by tobiasd** [comment]
> can you please add this repo to https://travis-ci.org for testing 

I'll have a look for this later.

File: develop/fpm-alpine/Dockerfile (hunk 8 above)
Comment (nupplaphil) [reply develop/fpm-alpine/Dockerfile]: How is Friendica normally using an email system? I know from the `dev` resources of the Friendica repository that you automatically install postfix and use it as a mail-server.

But how is it in production? For my environment, I do have a SMTP server for my local network where I'm routing the mails through.

One guess is that I'm automatically install `sendmail` and add "optional" SMTP-environment variables for the docker instances. What do you think?

**Comment by tobiasd** [comment]
> How is Friendica normally using an email system? I know from the
> `dev` resources of the Friendica repository that you automatically
> install postfix and use it as a mail-server.

Friendica uses the PHP mail command to send out emails. The server must provide the email functionality. There is an open feature request / idea to provide an alternative, but nothing that would work at the moment.

**Comment by nupplaphil** [comment]
I will give this solution a try: https://github.com/docker-library/php/issues/135#issuecomment-364392718

**Comment by nupplaphil** [comment]
I added two example-dockerfiles for customization:
- `cron` Combination of app & cron-job in one Docker-Image
- `smtp` Adding SMTP-capability for alpine and custom SMTP-Settings for ubuntu images

File: README.md [hunk 9]
```diff
@@ -1,3 +1,150 @@
...
+
+The apache image contains a webserver and exposes port 80. To start the container type:
+```console
+$ docker run -d -p 8080:80 --link some-mysql:mysql friendica 
+```
+
+Now you can access the Friendica installation wizard at http://localhost:8080/ from your host system.
+
+## Using the fpm image
+To use the fpm image you need an additional web server that can proxy http-request to the fpm-port of the container. For fpm connection this container exposes port 9000.
```
Comment (tobiasd) [inline README.md]: There are 2 sentences in the line.

Comment (nupplaphil) [reply README.md]: Ha, you got me :-)

**Comment by tobiasd** [comment]
I've added travis. It is currently running...

**Comment by nupplaphil** [comment]
Thank you @tobiasd :-)
As I said I do have problems with the `override-cmd` test of the official docker-repo test-suite.
I'm currently investigating what's wrong with it.

So please don't merge the PR yet.

I'll post if I'm ready :-)

**Comment by nupplaphil** [comment]
@tobiasd @MrPetovan 
Now I'm ready.

- `stable/apache` shows correctly the wizard
- `develop/apache` shows the "getCurrentTheme" exception ;-)
- I'm using `develop/fpm-alpine` in my environment with the `cron.sh` as well (I will push the whole environment later in a separated PR)

**Pull request merged**